### DISCUSSION TITLE
How to map encapsulated graph object as a relationship using imperative mapping

### DISCUSSION THREAD

**Discussion by ewolz3** [sqlalchemy/sqlalchemy#8452] [Usage Questions]
### Background
I have an object that uses an encapsulated graph object from the networkx library to store store state. A slimmed down version of my class looks something like:

```python
import uuid
import networkx as nx

class Hierarchy:
    def __init__(relationships=None):
        """
        While my class has other attributes, I'm only concerned with `self._graph` for this discussion
        """
        self.uuid = str(uuid.uuid4()
        self._graph = nx.DiGraph(relationships)

    def join(child_obj, parent_obj):
        ...
```

I want to map this graph object to a database table where each row represents an edge in the graph with 1 column being an identifier for the source node and a 2nd column being an identifier for the destination node. 

For example, if I have a graph object representing: child --> parent --> grandparent relationships, I would have a database table that looks like:
| source  | destination | hiearchy (fk) |
| ------------- | ------------- | ------------- |
| child  | parent  | abcd-efg-hijk-lmnop |
| parent  | grandparent  | abcd-efg-hijk-lmnop |

The graph object from networkx provides me with some views that would make this easier. For example, casting the `edges` property of the graph object to a list results in a list of tuples, where each tuple represents 1 edge in the graph and the elements being the node identifiers.
```python
>>> list(self._graph.edges)
[('child', 'parent'), ('parent', 'grandparent')]
```

### Questions on approach
If Hierarchy class represented the graph as a list of dataclasses, this would be fairly trivial to do using SqlAlchemy's relationship() and default list collection class. However, I really want to keep the internal state representation using the graph object. 

One approach I thought of was adding a `relationships` property/setter to my Hierarchy class that would transform the graph object to/from a list of dataclasses that I could then map to the database table using SqlAlchemy's relationship() with default list collection class. However, I run into issues mapping this `relationships` property because it is a descriptor. Not sure if I need to dig into SqlAlchemy [synonyms](https://docs.sqlalchemy.org/en/14/orm/mapped_attributes.html#synonyms) or [hybrid properties](https://docs.sqlalchemy.org/en/14/orm/mapped_attributes.html#using-descriptors-and-hybrids) to address this? My goal is for the Hierarchy object to have no knowledge of SqlAlchemy and I couldn't get this approach to work with my current understanding/proficiency level. 

I also started to read up on Custom Collection Classes [here](https://docs.sqlalchemy.org/en/14/orm/collections.html#custom-collection-implementations) which seemed to indicate that a custom collection class would be overkill and instead led me down the rabbit hole of using an Association Proxy?

> For the second use case, the [Association Proxy](https://docs.sqlalchemy.org/en/14/orm/extensions/associationproxy.html) extension is a well-tested, widely used system that provides a read/write “view” of a collection in terms of some attribute present on the target object. As the target attribute can be a @property that returns virtually anything, a wide array of “alternative” views of a collection can be constructed with just a few functions. This approach leaves the underlying mapped collection unaffected and avoids the need to carefully tailor collection behavior on a method-by-method basis.

This description sounded promising, but I'm getting completely lost in the weeds in the documentation for association proxy and can't figure out how to translate this to an implementation that solves my problem. 

### TL;DR
I think something in the documentation for the following topics is _probably_ what I need, but I can't seem to fully wrap my head around it:
- Custom collection class
- Association Proxy
- Synonyms
- Descriptors and Hybrid Property

**Comment by zzzeek** [comment]
> I want to map this graph object to a database table where each row represents an edge in the graph with 1 column being an identifier for the source node and a 2nd column being an identifier for the destination node.
> 
> For example, if I have a graph object representing: child --> parent --> grandparent relationships, I would have a database table that looks like:
> source 	destination 	hiearchy (fk)
> child 	parent 	abcd-efg-hijk-lmnop
> parent 	grandparent 	abcd-efg-hijk-lmnop

is "hierarchy" a column?  if so, that looks like three columns, not two ?   



> 
> The graph object from networkx provides me with some views that would make this easier. For example, casting the `edges` property of the graph object to a list results in a list of tuples, where each tuple represents 1 edge in the graph and the elements being the node identifiers.
> 
> ```python
> >>> list(self._graph.edges)
> [('child', 'parent'), ('parent', 'grandparent')]
> ```
> 
> ### Questions on approach
> 
> If Hierarchy class represented the graph as a list of dataclasses, this would be fairly trivial to do using SqlAlchemy's relationship() and default list collection class. However, I really want to keep the internal state representation using the graph object.
> 
> One approach I thought of was adding a `relationships` property/setter to my Hierarchy class that would transform the graph object to/from a list of dataclasses that I could then map to the database table using SqlAlchemy's relationship() with default list collection class. 

I can't give specific advice on techniques because I'm not understanding the model:

1. is Hierarchy a mapped class?  
2. is there an Edge class mapped to that table with the "source" / "destination" column?
3. do "source" / "destination" point to some other table?  or is this strictly a list of "Edges" with string names in them?
4. how does "networkx" fit into this?  are you trying to persist in the database a structure that you would read into a networkx structure when you load the object?
5. the relationship() is linking what exactly?

If I could get more of a POC of your model, without using assoc proxy / synontms any of that, then show me "here is the part I want to be more automatic and less exposed", we can work in the relevant APIs on a case by case basis.

**Reply by ewolz3** [reply]
Thanks for the reply, let me know if the following paints a clearer picture. 

> is Hierarchy a mapped class?

Yes. 

> is there an Edge class mapped to that table with the "source" / "destination" column?

Currently no (all the edge data is stored in the networkx graph object). 

The networkx graph object provides an `edges` view that can be iterated over to generate a list of tuples that representing the graph. Can these tuples be mapped to an Edges table or do I need to create my own view of the graph that returns a list of Edge classes (that would then be mapped to the edges table?)

> do "source" / "destination" point to some other table? or is this strictly a list of "Edges" with string names in them?

Yes, "source" and "destination" would be foreign keys pointing to some other table. For the sake of this example, lets call that table/class "Node" with a unique identifier "name" that "source" and "destination" would point to. 

> how does "networkx" fit into this? are you trying to persist in the database a structure that you would read into a networkx structure when you load the object?

Exactly. 

I'm struggling with where/how I transform this object into something that can be mapped to my Edges table. 

> If I could get more of a POC of your model, without using assoc proxy / synontms any of that, then show me "here is the part I want to be more automatic and less exposed", we can work in the relevant APIs on a case by case basis.

<details><summary>High level description</summary>
I've got a bunch of "Node" objects with some data. These Nodes can be related to each other, and this relationship is defined in my "Hierarchy" class. My goal is that "node" objects can be managed independently from the "hierarchy" object. If any "node" object gets deleted, it would cascade and delete any "edge" that references it from the hierarchy's graph.
</details>

<details><summary>Example class definitions</summary>

```python
class Node:
    def __init__(self, name: str, **kwargs):
        self.name = name
        # other fields here

class Hierarchy:
    """
    Manages relationships between nodes. The underlying graph object only maintains a 
    reference to the Node object using a unique identifier.
    """
    def __init__(self, relationships: List[Tuple]=None)
        self._graph = nx.DiGraph(relationships)
        # other fields here

    def join(child_obj: Node, parent_obj: Node):
        """
        Business logic to determine if a relationship between nodes can 
        be added. If so, add an edge to the graph with a reference to the 
        unique node identifier
        """
        self._graph.add_edge(child_obj.name, parent_obj.name)
```

</details>

<details><summary>Example usage</summary>

```python
child_node = Node('child')
parent_node = Node('parent')
gparent_node = Node('grandparent')
# To model relationships: child --> parent --> grandparent
hierarchy = Hierarchy()
hierarchy.join(child_node, parent_node)
hierarchy.join(parent_node, gparent_node)
# Shortcut to initializing the same hierarchy object is as follows:
# hierarchy = Hierarchy(relationships=[('child', 'parent'), ('parent', 'grandparent')])
```
</details>

<details><summary>Example database results</summary>

**Node**
| id  | name | Other data |
| ------------- | ------------- | ------------- |
| 1  | child  | ... |
| 2  | parent  | ... |
| 3  | grandparent  | ... |

**Hierarchy**
| id  | Other data|
| ------------- | ------------- |
| 1  | ... |

**Edges**
| source (fk to node.name) | destination (fk to node.name) | hierarchy_ref (fk to hierarchy.id) |
| ------------- | ------------- | ------------- |
| child  | parent | 1 |
| parent  | grandparentparent | 1 |
</details>

The mappings section in the following is where I believe I need help

```python
metadata = MetaData(naming_convention=...)
mapper_registry = registry(metadata=metadata)

# ------------------
# Table Definitions
# ------------------
nodes = Table(
    "nodes",
    mapper_registry.metadata,
    Column("id", Integer, primary_key=True, autoincrement=True),
    Column("name", String(255),
    # Other columns
)

hierarchy = Table(
    "hierarchy",
    mapper_registry.metadata,
    Column("id", Integer, primary_key=True, autoincrement=True),
    # Other columns
)

edges = Table(
    "edges",
    mapper_registry.metadata,
    Column("source", ForeignKey("nodes.name", ondelete="CASCADE"), primary_key=True), # Remove edge if source node is deleted
    Column("destination", ForeignKey("nodes.name", ondelete="CASCADE"), primary_key=True), # Remove edge if destination node is deleted
    Column("hierarchy_ref", ForeignKey("hierarchy.id", ondelete="CASCADE"), nullable=False)
)

# -----------
# Mappings
# -----------
mapper_registry.map_imperatively(models.Node, nodes)

edges_mapper = mapper_registry.map_imperatively(
    ??? <-- HELP HERE, can I map a tuple or do I need to create an Edge class?
    edges
)

hierarchy_mapper = mapper_registry.map_imperatively(
    models.Hierarchy,
    hierarchy,
    properties={
        "_graph": relationship( # <--- HELP HERE,  I know _graph is not right
            edges_mapper,
            cascade="all, delete-orphan", # if Hierarchy is deleted, all related edges in graph are deleted
            passive_deletes=True,
        )
    }
    # I want to transform the Hiearchy's `_graph` object into a representation that I can persist to the Edges table
    # and read back from the Edges table into my networkx graph object. The `_graph` object has an `edges` property
    # that can be iterated over to get a list of tuples (where each tuple represents an edge in the graph. First element in 
    # tuple represents the source node and the second element in the tuple represents the destination node)

    # The `__init__` of the networkx DiGraph object takes an optional list of tuples representing edges in the graph. 
    # Alternatively, I can create an emtpy DiGraph object and then populate it after the fact using the `add_edges_from()` 
    # method
)
```

**Reply by zzzeek** [reply]
you have to make an Edge class yes, since your edge table has three columns.    If it were nothing more than a simple association table, you could use [this pattern](https://docs.sqlalchemy.org/en/14/orm/join_conditions.html#self-referential-many-to-many-relationship) , but as each Edge is part of Hierarchy also, you need to use [association object](https://docs.sqlalchemy.org/en/14/orm/basic_relationships.html#association-object).

It should be simple enough to move between Edge objects and tuples of the two node columns, a simple list comprehension over hiearchy.edges should do that.

below is a POC for what you have so far.   

```py
from sqlalchemy import Column
from sqlalchemy import create_engine
from sqlalchemy import ForeignKey
from sqlalchemy import Integer
from sqlalchemy import MetaData
from sqlalchemy import String
from sqlalchemy import Table
from sqlalchemy.orm import registry
from sqlalchemy.orm import relationship
from sqlalchemy.orm import Session

metadata = MetaData()
mapper_registry = registry(metadata=metadata)

# ------------------
# Table Definitions
# ------------------

nodes = Table(
    "nodes",
    mapper_registry.metadata,
    Column("id", Integer, primary_key=True, autoincrement=True),
    Column("name", String(255)),
    # Other columns
)


hierarchy = Table(
    "hierarchy",
    mapper_registry.metadata,
    Column("id", Integer, primary_key=True, autoincrement=True),
    # Other columns
)

edges = Table(
    "edges",
    mapper_registry.metadata,
    Column(
        "source",
        ForeignKey("nodes.name", ondelete="CASCADE"),
        primary_key=True,
    ),  # Remove edge if source node is deleted
    Column(
        "destination",
        ForeignKey("nodes.name", ondelete="CASCADE"),
        primary_key=True,
    ),  # Remove edge if destination node is deleted
    Column(
        "hierarchy_ref",
        ForeignKey("hierarchy.id", ondelete="CASCADE"),
        nullable=False,
    ),
)


class Hierarchy:
    def edge_tuples(self):
        return [e.as_tuple() for e in self._graph]


class Node:
    pass


class Edge:
    def as_tuple(self):
        return (self.source, self.destination)


# -----------
# Mappings
# -----------
mapper_registry.map_imperatively(Node, nodes)

edges_mapper = mapper_registry.map_imperatively(
    Edge,
    edges,
    properties={
        "source_node": relationship(Node, foreign_keys=edges.c.source),
        "destination_node": relationship(
            Node, foreign_keys=edges.c.destination
        ),
    },
)

hierarchy_mapper = mapper_registry.map_imperatively(
    Hierarchy,
    hierarchy,
    properties={
        "_graph": relationship(
            edges_mapper,
            cascade="all, delete-orphan",
            passive_deletes=True,
        )
    },
)

e = create_engine("sqlite://", echo=True)
mapper_registry.metadata.create_all(e)

sess = Session(e)

n1, n2, n3 = Node(name="n1"), Node(name="n2"), Node(name="n3")
sess.add(
    Hierarchy(
        _graph=[
            Edge(source_node=n1, destination_node=n2),
            Edge(source_node=n2, destination_node=n3),
        ]
    )
)
sess.commit()


h1 = sess.get(Hierarchy, 1)
print(h1.edge_tuples())

```

**Reply by ewolz3** [reply]
Unless I'm missing something, I think this is still missing the marshalling to/from the networkx graph object.

For example, in
```python
n1, n2, n3 = Node(name="n1"), Node(name="n2"), Node(name="n3")
sess.add(
    Hierarchy(
        _graph=[
            Edge(source_node=n1, destination_node=n2),
            Edge(source_node=n2, destination_node=n3),
        ]
    )
)
sess.commit()
```
you have `_graph` as a list of the new Edge class, not a networkx DiGraph instance. 

Could the marshalling be done in a python property/setter somehow?

```python
class Edge:
    # <...snip...>
    def as_tuple():
         return (self.source, self.destination)

class Hierarchy:
    # <...snip...>
    @property
    def edge_list(self) -> List[Edge]:
        """Transforms networkx graph object into List of Edge objects"""
        return [Edge(edge_tuple[0], edge_tuple[1]) for edge_tuple in self._graph.edges]

    @edge_list.setter
    def edge_list(self, edges: List[Edge]):
        """Sets networkx graph object from List of Edge objects"""
        edge_tuples = [edge.as_tuple() for edge in edges]
        self._graph = nx.DiGraph(edge_tuples)
```

I'm still struggling with getting the mappings to work as I expect and suspect it has something to do with `hybrid_property` or `synonym`. 

```python
hierarchy_mapper = mapper_registry.map_imperatively(
    models.Hierarchy,
    hierarchy,
    properties={
        "edge_list": relationship( # <--- HELP HERE, this overwrites my property/setter with an SqlAlchemy `InstrumentedList`
            edges_mapper,
            cascade="all, delete-orphan",
            passive_deletes=True,
        ),
        "???": synonym(???, map_column=???, descriptor=models.Hierarchy.edge_list) # <-- HELP HERE at all the ???s
    }
```

[This ](https://docs.sqlalchemy.org/en/14/orm/extensions/hybrid.html#working-with-relationships) section of the documentation seems to describe something similar to what I'm looking for. However, with the classical/imperative mapping approach, I'm trying to keep my models from having any knowledge of SqlAlchemy (which means I don't want to define a `@hybrid_property` on my `Hierarchy` class). 

Since synonyms are a mapper level construct and are described as being able to do anything a hybrid_property can do, I'm hoping I'm just missing something simple.

**Reply by zzzeek** [reply]
> Unless I'm missing something, I think this is still missing the marshalling to/from the networkx graph object.
> 
> For example, in
> 
> ```python
> n1, n2, n3 = Node(name="n1"), Node(name="n2"), Node(name="n3")
> sess.add(
>     Hierarchy(
>         _graph=[
>             Edge(source_node=n1, destination_node=n2),
>             Edge(source_node=n2, destination_node=n3),
>         ]
>     )
> )
> sess.commit()
> ```
> 
> you have `_graph` as a list of the new Edge class, not a networkx DiGraph instance.

yes I didn't do anything related to that yet.


> 
> Could the marshalling be done in a python property/setter somehow?

sure!   that would be the usual way to do this.


> ```python
> class Edge:
>     # <...snip...>
>     def as_tuple():
>          return (self.source, self.destination)
> 
> class Hierarchy:
>     # <...snip...>
>     @property
>     def edge_list(self) -> List[Edge]:
>         """Transforms networkx graph object into List of Edge objects"""
>         return [Edge(edge_tuple[0], edge_tuple[1]) for edge_tuple in self._graph.edges]
> 
>     @edge_list.setter
>     def edge_list(self, edges: List[Edge]):
>         """Sets networkx graph object from List of Edge objects"""
>         edge_tuples = [edge.as_tuple() for edge in edges]
>         self._graph = nx.DiGraph(edge_tuples)
> ```
> 
> I'm still struggling with getting the mappings to work as I expect and suspect it has something to do with `hybrid_property` or `synonym`.

hybrid_property and synonym only have to do with a particular ORM-mapped attribute that one would like to be available as a SQL expression at the class level in some other way.   I would assume the DiGraph() object is something you only have at instance time.   synonyms and hybrids don't really enter the picture here.

also it seems like maybe you are thinking that the state of Hierarchy would be the DiGraph object, and that somehow at database persistence time it becomes an ORM-related list that...isn't stateful?   The list of Edge objects has to be part of the state of ...something if you want to use regular ORM persistence patterns.    Keeping them as a list on Hierarchy with relationship seems most straightforward.

the simplest way to have that DiGraph there would be that you make one on the fly from a single read-only `@property`, that would be if you wanted a view of edges in terms of a DiGraph object.   But if you want to make mutations to the DiGraph and have them be persisted, you would need to use event hooks.    There's a few ways to do that and it involves how you want mutations to the DiGraph to proceed.   Options to use here would include the [before_flush](https://docs.sqlalchemy.org/en/14/orm/events.html?highlight=before_flush#sqlalchemy.orm.SessionEvents.before_flush) event which could be used to rearrange the list of Edge objects to correspond to the DiGraph instance when the flush proceed, as well as [flag_modified](https://docs.sqlalchemy.org/en/14/orm/session_api.html#sqlalchemy.orm.attributes.flag_modified) when changes to the DiGraph occur.  But I would need to know more about what public interface to this DiGraph you had in mind.   If you wanted to allow direct access to the DiGraph object, or if you were going to have methods on Hierarchy that manipulate it internally (the latter would be easier).


> [This ](https://docs.sqlalchemy.org/en/14/orm/extensions/hybrid.html#working-with-relationships) section of the documentation seems to describe something similar to what I'm looking for. However, with the classical/imperative mapping approach, I'm trying to keep my models from having any knowledge of SqlAlchemy (which means I don't want to define a `@hybrid_property` on my `Hierarchy` class).

that can be done here also, although I think this kind of thing is not really pragmatic.    The ORM adds attributes to the class that implementations typically expect to exist.   Would your program be operating in some mode of use where SQLAlchemy were not installed at all, yet the Hierarchy object were still being manipulated and used ?   if so, I think it's important that you have those cases in your test suite because otherwise that kind of ideal is just theater, when the app in fact has no chance at all of running without SQLAlchemy ORM mappings actually being present.

regardless, that step would be secondary once we determine what interface you are looking to achieve.  


> 
> Since synonyms are a mapper level construct and are described as being able to do anything a hybrid_property can do, I'm hoping I'm just missing something simple.

**Reply by ewolz3** [reply]
Side note, I wanted to express my thanks for all the help/responses you've given so far!

That said

> also it seems like maybe you are thinking that the state of Hierarchy would be the DiGraph object, and that somehow at database persistence time it becomes an ORM-related list that...isn't stateful? The list of Edge objects has to be part of the state of ...something if you want to use regular ORM persistence patterns. Keeping them as a list on Hierarchy with relationship seems most straightforward.

I may not be understanding ORM persistence patterns correctly, but yes, I wanted to maintain the state of my `Hierarchy` class with the `DiGraph` object. I thought at database persistence time I could transform the DiGraph internal representation of its state (a "dictionary of dictionaries of dictionaries") to my own view (list of Edge objects) that would map to the desired database schemas. When reading from the database, I would read in a list of Edge objects and transform that to a list of tuples that could be used to hydrate the DiGraph object. If I load an object from the database, make changes, and then commit, the DiGraph internal representation of state would have changed, which would lead to a change in my own view of the object (list of Edges) which would then get saved. 

Is that not possible / am I thinking about this the totally wrong way?

> the simplest way to have that DiGraph there would be that you make one on the fly from a single read-only @property, that would be if you wanted a view of edges in terms of a DiGraph object. But if you want to make mutations to the DiGraph and have them be persisted, you would need to use event hooks. There's a few ways to do that and it involves how you want mutations to the DiGraph to proceed. Options to use here would include the [before_flush](https://docs.sqlalchemy.org/en/14/orm/events.html?highlight=before_flush#sqlalchemy.orm.SessionEvents.before_flush) event which could be used to rearrange the list of Edge objects to correspond to the DiGraph instance when the flush proceed, as well as [flag_modified](https://docs.sqlalchemy.org/en/14/orm/session_api.html#sqlalchemy.orm.attributes.flag_modified) when changes to the DiGraph occur.

If I'm understanding correctly, you are suggesting to instead maintain state on the object as a list of Edge objects instead of a networkx DiGraph object? And rather than have `_graph` be an attribute on the object, make it a `@property` that returns a DiGraph instance loaded from the list of Edges that are stored on the instance?

i.e.:
```python
class Hierarchy:
    def __init__(self, edges: List[Edge]=None):
        self._edges = edges or []

    @property
    def _graph(self):
        return nx.DiGraph([edge.as_tuple()] for edge in self._edges)
```

Makes sense for methods on the Hierarchy class that are read-only (which would be the majority), I'd have to think about it some more for my methods that change state of the DiGraph instance (adding edges or removing edges)

> But I would need to know more about what public interface to this DiGraph you had in mind. If you wanted to allow direct access to the DiGraph object, or if you were going to have methods on Hierarchy that manipulate it internally (the latter would be easier).

The way I have it set up right now, any interaction with the `_graph` attribute (DiGraph instance) of the `Hierarchy` class must be done through a public method or property. Users of the `Hierarchy` class should not need to be aware that we are using `networkx` under the hood. 

> that can be done here also, although I think this kind of thing is not really pragmatic. The ORM adds attributes to the class that implementations typically expect to exist. Would your program be operating in some mode of use where SQLAlchemy were not installed at all, yet the Hierarchy object were still being manipulated and used ? if so, I think it's important that you have those cases in your test suite because otherwise that kind of ideal is just theater, when the app in fact has no chance at all of running without SQLAlchemy ORM mappings actually being present.

Much of our existing test suite uses raw Python objects from our model without relying on the SqlAlchemy instrumentation. The goal was to separate concerns and reduce coupling between our business logic and the database layer.

**Reply by zzzeek** [reply]
> 
> I may not be understanding ORM persistence patterns correctly, but yes, I wanted to maintain the state of my `Hierarchy` class with the `DiGraph` object. I thought at database persistence time I could transform the DiGraph internal representation of its state (a "dictionary of dictionaries of dictionaries") to my own view (list of Edge objects) that would map to the desired database schemas. 
> When reading from the database, I would read in a list of Edge objects and transform that to a list of tuples that could be used to hydrate the DiGraph object. If I load an object from the database, make changes, and then commit, the DiGraph internal representation of state would have changed, which would lead to a change in my own view of the object (list of Edges) which would then get saved.

you could do that.  you could use the before_flush() event,  then you could just build the list of Edge objects and add them to the Session.  That side of it could be done in some other ways as well but that would be the simplest.     you'd also use the load() event to load up Edge objects and create the DiGraph object, and on that end, you get a lot of benefits by having a relationship on Hierarchy as it allows the Edge objects to be efficiently loaded.      you would need to also use the flag_modified() call to get the flush() to occur since Hierarchy is otherwise not going to be making ORM-level mutations by itself.

the potentially tricky part is calling that flag_modified() hook, which signals that your Hierarchy has pending changes to flush.  you would need to roll that part yourself, most simply by having mutator methods on Hierarchy that modify the DiGraph.

Another variant would be that Hierarchy would maintain DiGraph and Edge collections simultaneously but that seems like it would be more complicated.

> 
> > the simplest way to have that DiGraph there would be that you make one on the fly from a single read-only @Property, that would be if you wanted a view of edges in terms of a DiGraph object. But if you want to make mutations to the DiGraph and have them be persisted, you would need to use event hooks. There's a few ways to do that and it involves how you want mutations to the DiGraph to proceed. Options to use here would include the [before_flush](https://docs.sqlalchemy.org/en/14/orm/events.html?highlight=before_flush#sqlalchemy.orm.SessionEvents.before_flush) event which could be used to rearrange the list of Edge objects to correspond to the DiGraph instance when the flush proceed, as well as [flag_modified](https://docs.sqlalchemy.org/en/14/orm/session_api.html#sqlalchemy.orm.attributes.flag_modified) when changes to the DiGraph occur.
> 
> If I'm understanding correctly, you are suggesting to instead maintain state on the object as a list of Edge objects instead of a networkx DiGraph object? And rather than have `_graph` be an attribute on the object, make it a `@property` that returns a DiGraph instance loaded from the list of Edges that are stored on the instance?

that's one approach yes, but that's looking at DiGraph as more of a read-only view of things.  I also dont know how expensive it is to create, etc.   The event based approach above might be better for what you are doing.

> 
> Makes sense for methods on the Hierarchy class that are read-only (which would be the majority), I'd have to think about it some more for my methods that change state of the DiGraph instance (adding edges or removing edges)
> 
> > But I would need to know more about what public interface to this DiGraph you had in mind. If you wanted to allow direct access to the DiGraph object, or if you were going to have methods on Hierarchy that manipulate it internally (the latter would be easier).
> 
> The way I have it set up right now, any interaction with the `_graph` attribute (DiGraph instance) of the `Hierarchy` class must be done through a public method or property. Users of the `Hierarchy` class should not need to be aware that we are using `networkx` under the hood.

I figured that.


There is actually an example in the examples folder that does something like this.  It also uses some very old Python stuff, it's old.  Take  a look at the elementtree examples such as https://docs.sqlalchemy.org/en/14/_modules/examples/elementtree/adjacency_list.html where it maintains a list of DOM objects called _Node / _Attribute and moves them to/from an ElementTree structure.  This example uses a Python descriptor to do it (like a `@property` but explicit).    this approach could work for what you're doing also but it looks a little dated now.    When that example was made we didnt have all the event hooks we do now.

> 
> > that can be done here also, although I think this kind of thing is not really pragmatic. The ORM adds attributes to the class that implementations typically expect to exist. Would your program be operating in some mode of use where SQLAlchemy were not installed at all, yet the Hierarchy object were still being manipulated and used ? if so, I think it's important that you have those cases in your test suite because otherwise that kind of ideal is just theater, when the app in fact has no chance at all of running without SQLAlchemy ORM mappings actually being present.
> 
> Much of our existing test suite uses raw Python objects from our model without relying on the SqlAlchemy instrumentation. The goal was to separate concerns and reduce coupling between our business logic and the database layer.

right, I think if you really want that to be there, you should include in your tests use of your model objects without SQLAlchemy being present, otherwise you are still dependent on SQLAlchemy.  Like if you have a method that says, "if self.name == 'x'", if ".name" is there only because SQLAlchemy mapped a descriptor there, you now have a functional dependency on the ORM that is distinct from the concern of database persistence.    The entire reason SQLAlchemy has "imperative mappings" is that back in 2006, I too thought that was the best way to do things, that you can have your business classes and the "mapper" would be completely separate.    But the ORM applies instrumentation to attributes on the class.  The class itself then becomes written to assume the presence of these attributes as well as their specific behavior which is very idiosyncratic to SQLAlchemy, such as, attributes that aren't defined are returned as "None" or even an empty list, there are ORM features like attribute validation, there's backrefs, deferred loading, etc.   That was when I realized that imperative mappings are not really delivering anything in practice and I was even going to remove the whole thing, but people complained.

anyway here is the POC with before_flush / load implemented

```py
from sqlalchemy import Column
from sqlalchemy import create_engine
from sqlalchemy import ForeignKey
from sqlalchemy import Integer
from sqlalchemy import MetaData
from sqlalchemy import String
from sqlalchemy import Table, select, tuple_
from sqlalchemy.orm import registry
from sqlalchemy.orm.attributes import flag_modified
from sqlalchemy.orm import relationship
from sqlalchemy.orm import sessionmaker
from sqlalchemy import event

metadata = MetaData()
mapper_registry = registry(metadata=metadata)

# ------------------
# Table Definitions
# ------------------

nodes = Table(
    "nodes",
    mapper_registry.metadata,
    Column("id", Integer, primary_key=True, autoincrement=True),
    Column("name", String(255)),
    # Other columns
)


hierarchy = Table(
    "hierarchy",
    mapper_registry.metadata,
    Column("id", Integer, primary_key=True, autoincrement=True),
    # Other columns
)

edges = Table(
    "edges",
    mapper_registry.metadata,
    Column(
        "source",
        ForeignKey("nodes.name", ondelete="CASCADE"),
        primary_key=True,
    ),  # Remove edge if source node is deleted
    Column(
        "destination",
        ForeignKey("nodes.name", ondelete="CASCADE"),
        primary_key=True,
    ),  # Remove edge if destination node is deleted
    Column(
        "hierarchy_ref",
        ForeignKey("hierarchy.id", ondelete="CASCADE"),
        nullable=False,
    ),
)


class DiGraph:
    def __init__(self, tuples):
        self.tuples = tuples

    def append(self, edge):
        self.tuples.append(edge)

    def __str__(self):
        return f"DiGraph({self.tuples})"


class Hierarchy:
    def __init__(self, edges):
        self._graph = DiGraph(edges)

    def append_edge(self, edge):
        # ORM - adds a note to Hierarchy._sa_instance_state that will include
        # the object in the next flush.
        # this is a persistence detail for which you'd need to use some
        # creativity, like a decorator or some kind of abstract "update_state"
        # method, in order to hide it from being exposed in Hierarchy
        flag_modified(self, "_edge_persistence")
        self._graph.append(edge)


class Node:
    pass


class Edge:
    pass


# -----------
# Mappings
# -----------
mapper_registry.map_imperatively(Node, nodes)

edges_mapper = mapper_registry.map_imperatively(
    Edge,
    edges,
    properties={
        "source_node": relationship(Node, foreign_keys=edges.c.source),
        "destination_node": relationship(
            Node, foreign_keys=edges.c.destination
        ),
    },
)

hierarchy_mapper = mapper_registry.map_imperatively(
    Hierarchy,
    hierarchy,
    properties={
        "_edge_persistence": relationship(
            edges_mapper,
            cascade="all, delete-orphan",
            passive_deletes=True,
        )
    },
)


e = create_engine("sqlite://", echo=True)
mapper_registry.metadata.create_all(e)


Session = sessionmaker(bind=e)


@event.listens_for(Hierarchy, "load")
@event.listens_for(Hierarchy, "refresh")
def _load_digraphs(target, context, attrs=None):
    # populate DiGraph on load / refresh
    target._graph = DiGraph(
        [(e.source, e.destination) for e in target._edge_persistence]
    )


@event.listens_for(Session, "before_flush")
def _persist_digraphs(session, flush_ctx, objects):

    # persist DiGraph on flush
    for obj in session.new | session.dirty:
        if not isinstance(obj, Hierarchy):
            continue

        # Normally I'd make this a method on Hierarchy, but we are trying
        # to keep persistence details away from Hierarchy.

        # Edge objects have primary key etc. so we have to merge the new
        # state in terms of the edges that are there.  you could also
        # delete / insert rows every time but that would be wasteful
        all_edges = {
            (Edge.source, Edge.destination): Edge
            for Edge in obj._edge_persistence
        }

        def _get_edge(source, dest):
            if (source, dest) in all_edges:
                # use Edge we already have
                return all_edges[(source, dest)]
            else:
                # make new Edge for one we don't
                all_edges[(source, dest)] = edge = Edge(
                    source=source, destination=dest
                )
                return edge

        # replace the collection.  Edge objects that aren't here anymore
        # will get deleted due to delete-orphan
        obj._edge_persistence = [
            _get_edge(source, dest) for source, dest in obj._graph.tuples
        ]


sess = Session()

n1, n2, n3, n4 = (
    Node(name="n1"),
    Node(name="n2"),
    Node(name="n3"),
    Node(name="n4"),
)
sess.add_all([n1, n2, n3, n4])

h1 = Hierarchy([("n1", "n2"), ("n2", "n3")])

sess.add(h1)
sess.commit()

h1 = sess.get(Hierarchy, 1)

h1.append_edge(("n3", "n4"))

sess.commit()
sess.close()

h1 = sess.get(Hierarchy, 1)

print(h1._graph)

```
